File: tutorial/05-arrays/problems/merge-sort.py
class MergeSort:
    def sort(self,arr,left,right):
        if left < right:
            mid= left+ (right-left)//2
            self.sort(arr,left,mid)
            self.sort(arr,mid+1,right)
            self.merge(arr,left,mid,right)
    def merge(self,arr,left,mid,right):
        i=j=0
        current=left
        temp1= arr[left:(mid+1)]
        temp2= arr[(mid+1):(right+1)]
        temp1Length=len(temp1)
        temp2Length=len(temp2)
        print('first',arr)
        while i < temp1Length and j < temp2Length : 
            if temp1[i] < temp2[j]:
                arr[current]=temp1[i]
                i+=1
            else :
                arr[current]=temp2[j]
                j+=1
            current+=1
        while i < temp1Length:
            arr[current]=temp1[i]
            i+=1
            current+=1
        while j < temp2Length:
            arr[current]=temp2[j]
            j+=1
            current+=1
        print('second',arr)
soft = MergeSort()
arr=[4,2,5,9,1]
soft.sort(arr,0,len(arr)-1)
print(arr)
        

File: tutorial/05-arrays/problems/merge-sort.ts
class MergeSort {
  private _sort(arr: number[], left: number, right: number) {
    if (left < right) {
      const mid = left + Math.floor((right - left) / 2);
      this._sort(arr, left, mid);
      this._sort(arr, mid + 1, right);
      this._merge(arr, left, mid, right);
    }
  }
  private _merge(arr: number[], left: number, mid: number, right: number) {
    let i = 0,
      j = 0,
      current = left;
    const temp1 = arr.slice(left, mid + 1);
    const temp2 = arr.slice(mid + 1, right + 1);
    const temp1Length = temp1.length;
    const temp2Length = temp2.length;
    while (i < temp1Length && j < temp2Length) {
      if (temp1[i] < temp2[j]) {
        arr[current] = temp1[i];
        i++;
      } else {
        arr[current] = temp2[j];
        j++;
      }
      current++;
    }
    while (i < temp1Length) {
      arr[current] = temp1[i];
      i++;
      current++;
    }
    while (j < temp2Length) {
      arr[current] = temp2[j];
      j++;
      current++;
    }
  }
  public sort(arr: number[]) {
    this._sort(arr, 0, arr.length - 1);
  }
}
const soft = new MergeSort();
const arr = [4, 3, 2, 1, 7, 8];
soft.sort(arr);
console.log(arr);
